src/applets/caffeine/CaffeineWindow.vala

```diff
@@ -0,0 +1,188 @@
+        try {
+            props = Bus.get_proxy_sync (BusType.SESSION,
+                                        "org.gnome.SettingsDaemon.Power",
+                                        "/org/gnome/SettingsDaemon/Power");
+        } catch (IOError e) {
+            print ("Error: %s\n", e.message);
+        }
+        fetch_default ();
+
+        mode.notify["active"].connect (on_mode_active);
+    }
+
+    public void toggle_applet ()
+    {
+        if (mode.active) {
+            mode.active = false;
+            return;
+        }
+        mode.active = true;
+    }
+
+    private void fetch_default ()
+    {
+        // Fetch default configuration
+        default_idle_delay = session_settings.get_uint ("idle-delay");
+        default_idle_dim = power_settings.get_boolean ("idle-dim");
+        default_sleep_inactive_ac_type = power_settings.get_string ("sleep-inactive-ac-type");
+        default_sleep_inactive_battery_type = power_settings.get_string ("sleep-inactive-battery-type");
+        try {
+            default_brightness = props.brightness;
+        } catch (DBusError e) {
+            print ("Error: %s\n", e.message);
+        }
+    }
+
+    private void change_brightness (int32 value)
+    {
+        try {
+            props.brightness = value;
+        } catch (DBusError e) {
+            print ("Error: %s\n", e.message);
+        }
+    }
+
+    private void send_notification (bool activate, int time)
+    {
+        var cmd = new StringBuilder ();
+        cmd.append ("notify-send ");
+
+        if (activate) {
+            cmd.append ("\"%s\" ".printf (_("Turn on Caffeine Boost")));
+
+            if (time > 0) {
+                var duration = ngettext ("a minute", "%d minutes", time).printf (time);
+                cmd.append ("\"%s %s\" ".printf (_("Will turn off in"), duration));
+            }
+            cmd.append ("--icon=caffeine-cup-full");
+        } else {
+            cmd.append ("\"%s\" ".printf (_("Turn off Caffeine Boost")));
+            cmd.append ("--icon=caffeine-cup-empty");
+        }
+
+        try {
+            Process.spawn_command_line_async (cmd.str);
+        } catch (SpawnError e) {
+            print ("Error: %s\n", e.message);
+        }
+    }
+
+    private void on_mode_active (Object? obj, ParamSpec? params)
+    {
+        var icon = event_box.get_child ();
+        event_box.remove (icon);
+
+        var time = timer.get_value_as_int();
+        if (mode.active)
+        {
+            // Fetch power settings default value
+            fetch_default ();
+
+            timer.sensitive = false;
+            session_settings.set_uint ("idle-delay", 0);
+            power_settings.set_boolean ("idle-dim", false);
+            power_settings.set_string ("sleep-inactive-ac-type", "nothing");
+            power_settings.set_string ("sleep-inactive-battery-type", "nothing");
+
+            icon = new Gtk.Image.from_icon_name("caffeine-cup-full", Gtk.IconSize.MENU);
```

Comment:
icon setting can use a ternary operator on the string, along the lines of (but not necessarily representative of working code :D):

```
icon = new Gtk.Image.from_icon_name("caffeine-cup-" + (mode.active ? "full" : "empty"), Gtk.IconSize.MENU);
```

This and the event_box add call could be done outside of the if/else block.